File: web/ops-central-ui/src/services/dashboardService.ts
// Dashboard API Service
const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8080';

export interface DashboardMetrics {
    // Scheduler Metrics
    queue_depth: number;
    active_tasks: number;
    max_concurrency: number;
    worker_saturation: number;
    circuit_breaker_state: string;
    admission_mode: string;
    runtime_mode: string;

    // Leadership Metrics
    is_leader: boolean;
    current_epoch: number;
    leader_transitions: number;
    node_id: string;

    // Store Metrics
    pending_states: number;
    drifted_states: number;
    active_agents: number;

    // Multi-Cluster
    cluster_id: string;
    cluster_role: string;
    region: string;

    // Timestamp
    timestamp: number;
}

export interface ClusterInfo {
    cluster_id: string;
    region: string;
    role: string;
    is_leader: boolean;
    agent_count: number;
    health_score: number;
    endpoint: string;
}

export interface IncidentSnapshot {
    incident_id: string;
    state_id: string;
    node_id: string;
    failure_reason: string;
    timestamp: number;
    scheduler_snapshot: {
        queue_depth: number;
        active_tasks: number;
        worker_saturation: number;
        circuit_breaker_state: string;
        runtime_mode: string;
    };
    leader_snapshot: {
        is_leader: boolean;
        current_epoch: number;
        node_id: string;
    };
    timeline: Array<{
        req_id: string;
        stage: string;
        node_id: string;
        tenant_id: string;
        timestamp: number;
    }>;
}

class DashboardService {
    private baseUrl: string;

    constructor(baseUrl: string = API_BASE_URL) {
        this.baseUrl = baseUrl;
    }

    async getDashboardMetrics(): Promise<DashboardMetrics> {
        const response = await fetch(`${this.baseUrl}/api/dashboard`);
        if (!response.ok) {
            throw new Error(`Failed to fetch dashboard metrics: ${response.statusText}`);
        }
        return response.json();
    }

    async getClusters(): Promise<ClusterInfo[]> {
        const response = await fetch(`${this.baseUrl}/api/clusters`);
        if (!response.ok) {
            throw new Error(`Failed to fetch clusters: ${response.statusText}`);
        }
        return response.json();
    }

    async getIncidents(): Promise<IncidentSnapshot[]> {
        const response = await fetch(`${this.baseUrl}/api/incidents`);
        if (!response.ok) {
            throw new Error(`Failed to fetch incidents: ${response.statusText}`);
        }
        return response.json();
    }

    async replayIncident(incidentId: string): Promise<any> {
        const response = await fetch(`${this.baseUrl}/api/incidents/replay/${incidentId}`, {
            method: 'POST',
        });
        if (!response.ok) {
            throw new Error(`Failed to replay incident: ${response.statusText}`);
        }
        return response.json();
    }

    async captureIncidentSnapshot(stateId: string): Promise<IncidentSnapshot> {
        const response = await fetch(`${this.baseUrl}/api/incidents/capture?state_id=${stateId}`, {
            method: 'POST',
        });
        if (!response.ok) {
            throw new Error(`Failed to capture incident: ${response.statusText}`);
        }
        return response.json();
    }

    createWebSocketConnection(): WebSocket {
        const wsUrl = this.baseUrl.replace('http', 'ws');
        return new WebSocket(`${wsUrl}/api/dashboard/stream`);
    }
}

export const dashboardService = new DashboardService();

File: control_plane/dashboard_service.go
package main

import (
	"context"
	"time"

	"github.com/itskum47/FluxForge/control_plane/coordination"
	"github.com/itskum47/FluxForge/control_plane/scheduler"
	"github.com/itskum47/FluxForge/control_plane/store"
)

// DashboardService provides an abstraction layer for dashboard data retrieval.
// It decouples the API from direct store access and aggregates data from multiple sources.
type DashboardService struct {
	store     store.Store
	scheduler *scheduler.Scheduler
	elector   *coordination.LeaderElector
}

// NewDashboardService creates a new DashboardService.
func NewDashboardService(store store.Store, scheduler *scheduler.Scheduler, elector *coordination.LeaderElector) *DashboardService {
	return &DashboardService{
		store:     store,
		scheduler: scheduler,
		elector:   elector,
	}
}

// GetDashboardMetrics collects and aggregates all metrics for a specific tenant.
func (s *DashboardService) GetDashboardMetrics(ctx context.Context, tenantID string) (DashboardMetrics, error) {
	// 1. Scheduler Metrics
	schedMetrics := s.scheduler.GetMetrics()

	// 2. Leadership Metrics
	var leaderState coordination.LeaderState
	if s.elector != nil {
		leaderState = s.elector.GetState()
	}

	// 3. Store Metrics (Tenant Scoped)
	pending, err := s.store.CountStatesByStatus(ctx, tenantID, "pending")
	if err != nil {
		return DashboardMetrics{}, err
	}

	drifted, err := s.store.CountStatesByStatus(ctx, tenantID, "drifted")
	if err != nil {
		return DashboardMetrics{}, err
	}

	agents, err := s.store.ListAgents(ctx, tenantID)
	if err != nil {
		return DashboardMetrics{}, err
	}

	// 4. Construct Response
	return DashboardMetrics{
		// Scheduler
		QueueDepth:          schedMetrics.QueueDepth,
		ActiveTasks:         schedMetrics.ActiveTasks,
		MaxConcurrency:      schedMetrics.MaxConcurrency,
		WorkerSaturation:    schedMetrics.WorkerSaturation,
		CircuitBreakerState: schedMetrics.CircuitBreakerState,
		AdmissionMode:       schedMetrics.AdmissionMode,
		RuntimeMode:         schedMetrics.RuntimeMode,

		// Leadership
		IsLeader:          leaderState.IsLeader,
		CurrentEpoch:      leaderState.CurrentEpoch,
		LeaderTransitions: leaderState.Transitions,
		NodeID:            leaderState.NodeID,

		// Store
		PendingStates: pending,
		DriftedStates: drifted,
		ActiveAgents:  len(agents),

		// Cluster Info
		ClusterID:   "cluster-primary",
		ClusterRole: getClusterRole(leaderState.IsLeader),
		Region:      "us-east-1",
		Timestamp:   time.Now().Unix(),
	}, nil
}

func getClusterRole(isLeader bool) string {
	if isLeader {
		return "leader"
	}
	return "follower"
}
